source/base/tensor_product_polynomials.cc

```diff
@@ -303,6 +303,192 @@ TensorProductPolynomials<0, Polynomials::Polynomial<double>>::compute_grad_grad(
 
 
 
+namespace internal
+{
+  namespace TensorProductPolynomials
+  {
+    // This function computes the tensor product of some tabulated
+    // one-dimensional polynomials (also the anisotropic case is supported)
+    // with tensor product indices of all dimensions except the first one
+    // tabulated in the 'indices' array; the first dimension is manually
+    // iterated through because these are possibly performance-critical loops,
+    // so we want to avoid indirect addressing.
+    template <int dim, std::size_t dim1>
+    void
+    evaluate_tensor_product(
+      const unsigned int n_derivatives,
+      const boost::container::small_vector<dealii::ndarray<double, 5, dim>, 10>
+        &                values_1d,
+      const unsigned int size_x,
+      const boost::container::small_vector<std::array<unsigned int, dim1>, 64>
+        &                              indices,
+      const std::vector<unsigned int> &index_map,
+      std::vector<double> &            values,
+      std::vector<Tensor<1, dim>> &    grads,
+      std::vector<Tensor<2, dim>> &    grad_grads,
+      std::vector<Tensor<3, dim>> &    third_derivatives,
+      std::vector<Tensor<4, dim>> &    fourth_derivatives)
+    {
+      const bool update_values = (values.size() == indices.size() * size_x),
+                 update_grads  = (grads.size() == indices.size() * size_x),
+                 update_grad_grads =
+                   (grad_grads.size() == indices.size() * size_x),
+                 update_3rd_derivatives =
+                   (third_derivatives.size() == indices.size() * size_x),
+                 update_4th_derivatives =
+                   (fourth_derivatives.size() == indices.size() * size_x);
```

Comment:
I think we avoid using this style.
```suggestion
      const bool update_values = (values.size() == indices.size() * size_x);
      const bool update_grads  = (grads.size() == indices.size() * size_x);
      const bool update_grad_grads =
                   (grad_grads.size() == indices.size() * size_x);
      const bool update_3rd_derivatives =
                   (third_derivatives.size() == indices.size() * size_x);
      const bool update_4th_derivatives =
                   (fourth_derivatives.size() == indices.size() * size_x);
```